cmd/tools/testwick/testwicker.go

```diff
@@ -0,0 +1,246 @@
+		w.logger.WithError(w.Error).Error()
+		return w
+	}
+	w.channelID = channel.Id
+	return w
+}
+
+// CreateIncomingWebhook creates an incoming webhook
+func (w *TestWicker) CreateIncomingWebhook() *TestWicker {
+	w.logger.WithField("DNS", w.installation.DNS).Info("Creating incoming webhook")
+	if w.userID == "" {
+		w.Error = fmt.Errorf("failed to post message. You need to create a user first")
+		return w
+	}
+	if w.channelID == "" {
+		w.Error = fmt.Errorf("failed to post message. You need to create channels first")
+		return w
+	}
+	_, response := w.mmClient.CreateIncomingWebhook(&mmodel.IncomingWebhook{
+		ChannelId: w.channelID,
+		UserId:    w.userID,
+	})
+	if response.StatusCode != 201 {
+		w.Error = fmt.Errorf("failed to create channel status = %d, message = %s", response.StatusCode, response.Error.Message)
+		w.logger.WithError(w.Error).Error()
+		return w
+	}
+	return w
+}
+
+// WaitForInstallationStable waits for installation to become stable
+func (w *TestWicker) WaitForInstallationStable(ctx context.Context) *TestWicker {
+	w.logger.WithField("DNS", w.installation.DNS).Info("Waiting installation")
+	for {
+		i, err := w.provisionerClient.GetInstallation(w.installation.ID, &cmodel.GetInstallationRequest{})
+		if err != nil {
+			w.Error = errors.Wrap(err, "client.GetInstallation")
+			return w
+		}
+		w.logger.WithFields(logrus.Fields{
+			"DNS":   w.installation.DNS,
+			"State": w.installation.State,
+		}).Info("Waiting installation")
+
+		if i.State == cmodel.InstallationStateStable {
+			w.installation = i
+			return w
+		}
+		if i.State == cmodel.InstallationStateCreationFailed {
+			w.installation = i
+			w.Error = errors.Wrapf(err, "Installation creation failed with ID: %s", w.installation.ID)
+			return w
+		}
+
+		select {
+		case <-ctx.Done():
+			w.Error = errors.New("timed out waiting to become stable")
+			return w
+		case <-time.After(10 * time.Second):
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// SetupInstallation creates a user and login with this one
+func (w *TestWicker) SetupInstallation() *TestWicker {
+	u := &mmodel.User{
+		Username: "testwick",
+		Email:    "testwick@example.mattermost.com",
+		Password: "T3stw1ck10!",
+	}
+	_, response := w.mmClient.CreateUser(u)
+	if response.StatusCode != 201 {
+		w.Error = fmt.Errorf("failed to create admin user status = %d, message = %s", response.StatusCode, response.Error.Message)
+		return w
+	}
+	w.mmClient.Logout()
+	userLogged, response := w.mmClient.Login(u.Username, u.Password)
+	if response.StatusCode != 200 {
+		w.Error = fmt.Errorf("failed logging user: username = %s, status code = %d, message = %s", u.Username, response.StatusCode, response.Error.Message)
+		return w
+	}
+	w.userID = userLogged.Id
+	return w
+}
+
+// IsUP cpings the installation till return statusCode=200 and status=OK
+func (w *TestWicker) IsUP(ctx context.Context) *TestWicker {
+	w.mmClient = mmodel.NewAPIv4Client(fmt.Sprintf("https://%s", w.installation.DNS))
```

Comment:
Could we initialize the client earlier? I think we already have a DNS prior to the creation of installation, and initializing it here might lead to some confusion/errors if we will be extending the code, adding new methods, and mixing things up.

Reply:
sure agree